src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -63,23 +66,51 @@ public LocalLockManager getLocalLockManager()
     }
 
     @Override
-    public void updateDominantLocks( String path, LockLevel level )
-                    throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
-                    RollbackException
+    public void updateDominantLocks( String path, UnlockStatus unlockStatus ) throws PartylineException
     {
+        // Do nothing if dominance did not change
+        if ( !unlockStatus.dominanceChanged )
+        {
+            return;
+        }
         TransactionManager transactionManager = metadataCache.getAdvancedCache().getTransactionManager();
-        transactionManager.begin();
-        FileMeta meta = metadataCache.get( path );
-        if ( level == null )
+        try
         {
-            meta.removeLock( this.nodeKey );
+            transactionManager.begin();
+            FileMeta meta = metadataCache.get( path );
+            if ( unlockStatus.dominantLockLevel == null )
+            {
+                meta.removeLock( this.nodeKey );
+            }
+            else
+            {
+                meta.addLock( this.nodeKey, unlockStatus.dominantLockLevel );
+            }
+            metadataCache.put( path, meta );
+        }
+        catch ( NotSupportedException | SystemException e )
+        {
+            try
+            {
+                transactionManager.rollback();
+            }
+            catch ( SystemException ex )
+            {
+                throw new PartylineException( "There was a problem with a transaction rollback involving " + path );
```

Comment:
This has to turn into a log statement, since it will cause the original exception to be lost.